kymatio/scattering2d/backend/tensorflow_backend.py

```diff
@@ -0,0 +1,176 @@
+
+    def __call__(self, x):
+        if self.pre_pad:
+            return x
+        else:
+            paddings = [[0, 0]]*len(x.shape[:-2])
+            paddings += [[self.pad_size[0],self.pad_size[1]],[self.pad_size[2],self.pad_size[3]]]
+            return tf.cast(tf.pad(x, paddings, mode="REFLECT"), tf.complex64)
+
+def unpad(in_):
+    """
+        Slices the input tensor at indices between 1::-1
+        Parameters
+        ----------
+        in_ : tensor_like
+            input tensor
+        Returns
+        -------
+        in_[..., 1:-1, 1:-1]
+    """
+    return in_[..., 1:-1, 1:-1]
+
+class SubsampleFourier(object):
+    """
+        Subsampling of a 2D image performed in the Fourier domain
+        Subsampling in the spatial domain amounts to periodization
+        in the Fourier domain, hence the formula.
+        Parameters
+        ----------
+        x : tensor_like
+            input tensor with at least 5 dimensions, the last being the real
+             and imaginary parts.
+            Ideally, the last dimension should be a power of 2 to avoid errors.
+        k : int
+            integer such that x is subsampled by 2**k along the spatial variables.
+        Returns
+        -------
+        res : tensor_like
+            tensor such that its fourier transform is the Fourier
+            transform of a subsampled version of x, i.e. in
+            FFT^{-1}(res)[u1, u2] = FFT^{-1}(x)[u1 * (2**k), u2 * (2**k)]
+    """
+    def __call__(self, x, k):
+        y = tf.reshape(x,(-1, k, x.shape[1] // k, k, x.shape[2] // k))
+        
+        out = tf.reduce_mean(y, axis=(1, 3))
+        return out
+
+
+class Modulus(object):
+    """
+        This class implements a modulus transform for complex numbers.
+        Usage
+        -----
+        modulus = Modulus()
+        x_mod = modulus(x)
+        Parameters
+        ---------
+        x: input complex tensor.
+        Returns
+        -------
+        output: a real tensor equal to the modulus of x.
+    """
+    def __call__(self, x):
+        norm = tf.abs(x)
+        return tf.cast(norm, tf.complex64)
+
+
+def fft(x, direction='C2C', inverse=False):
+    """
+        Interface with torch FFT routines for 2D signals.
+        Example
+        -------
+        x = torch.randn(128, 32, 32, 2)
+        x_fft = fft(x, inverse=True)
+        Parameters
+        ----------
+        input : tensor
+            complex input for the FFT
+        direction : string
+            'C2R' for complex to real, 'C2C' for complex to complex
+        inverse : bool
+            True for computing the inverse FFT.
+            NB : if direction is equal to 'C2R', then an error is raised.
+    """
+    if direction == 'C2R':
+        if not inverse:
+            raise RuntimeError('C2R mode can only be done with an inverse FFT.')
+
+    if direction == 'C2R':
+        output = tf.math.real(tf.signal.ifft2d(x, name='irfft2d'))*tf.cast(x.shape[-1]*x.shape[-2],tf.float32)
```

Comment:
Will this normalization be needed after #470 ?

Reply:
I don't think so. But for now, it is necessary. I'll update the second paper merged such that it'llwork.

Reply:
Agreed.

Reply:
should be fine.